patches/server/0010-Adventure.patch

```diff
@@ -1213,6 +1039,72 @@ index 0000000000000000000000000000000000000000..d496a9a6ad229e42f1c44e31eafa6974
 +        return LegacyComponentSerializer.legacySection().serialize(player.adventure$displayName);
 +    }
 +}
+diff --git a/src/main/java/io/papermc/paper/adventure/ImprovedChatDecorator.java b/src/main/java/io/papermc/paper/adventure/ImprovedChatDecorator.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..f6005e656f4ca985de57b7c9931fffa13cb61ef4
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/adventure/ImprovedChatDecorator.java
+@@ -0,0 +1,60 @@
++package io.papermc.paper.adventure;
++
++import io.papermc.paper.event.player.AsyncChatCommandDecorateEvent;
++import io.papermc.paper.event.player.AsyncChatDecorateEvent;
++import net.minecraft.commands.CommandSourceStack;
++import net.minecraft.network.chat.ChatDecorator;
++import net.minecraft.server.MinecraftServer;
++import net.minecraft.network.chat.Component;
++import net.minecraft.server.level.ServerPlayer;
++import org.bukkit.craftbukkit.entity.CraftPlayer;
++import org.checkerframework.checker.nullness.qual.NonNull;
++import org.checkerframework.framework.qual.DefaultQualifier;
++
++import javax.annotation.Nullable;
++import java.util.concurrent.CompletableFuture;
++
++@DefaultQualifier(NonNull.class)
++public final class ImprovedChatDecorator implements ChatDecorator {
++    private final MinecraftServer server;
++
++    public ImprovedChatDecorator(MinecraftServer server) {
++        this.server = server;
++    }
++
++    @Override
++    public CompletableFuture<Component> decorate(@Nullable ServerPlayer sender, Component message) {
++        return decorate(server, sender, null, message);
++    }
++
++    @Override
++    public CompletableFuture<Component> decorate(@Nullable ServerPlayer sender, @Nullable CommandSourceStack commandSourceStack, Component message) {
++        return decorate(server, sender, commandSourceStack, message);
++    }
++
++    private static CompletableFuture<Component> decorate(final MinecraftServer server, final @Nullable ServerPlayer player, final @Nullable CommandSourceStack commandSourceStack, final Component originalMessage) {
++        return CompletableFuture.supplyAsync(() -> {
++            // Make a copy so that if the event if cancelled we can guarantee the original is being returned.
++            final net.kyori.adventure.text.Component originalMessageCopy = PaperAdventure.asAdventure(originalMessage.copy());
++            final net.kyori.adventure.text.Component initialResult = PaperAdventure.asAdventure(originalMessage.copy());
```

Comment:
These two calls to `.copy()` are not needed. Calling `PaperAdventure#asAdventure` already copies it. In addition, adventure's components are immutable, so you don't need to create 2 of these, just pass the same instance in for both params of the event.